### PR TITLE
Fix Rust sysroot

### PR DESCRIPTION
**Pull request by astro**
###### Description of changes

- Fixes a call by passing the required `lib` argument
- Update the the sysroot-src crate's Cargo.lock to the current rustc
- Makes the sysroot-src crate `#![no_std]` so that it does not attempt to build std
  - targets with std are fetched as binaries instead of being built by the sysroot, right?

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/build-support/rust/sysroot/src.nix [hunk 1]
```diff
@@ -19,7 +19,7 @@ stdenv.mkDerivation {
   + ''
     ${buildPackages.python3.withPackages (ps: with ps; [ toml ])}/bin/python3 ${./cargo.py}
     mkdir -p $out/src
-    touch $out/src/lib.rs
+    echo '#! [no_std]' > $out/src/lib.rs
```
Comment (Ericson2314) [inline pkgs/build-support/rust/sysroot/src.nix:22]:
```suggestion
    echo '#![no_std]' > $out/src/lib.rs
```
would be more idiomatic, right?

**Comment by winterqt** [comment]
I've marked this as a draft per the title.

**Comment by winterqt** [comment]
Also: please fill in the PR template (specifically motivation/description and testing) where possible, as well as add some descriptions to the commit messages.

Oh, lastly: why is this a draft, does it not work/does more need to be done to get this up to shape?

I will do a proper review of this, but I'd appreciate it if those points were addressed in the meantime.

Thank you!

**Comment by astro** [comment]
> Oh, lastly: why is this a draft, does it not work/does more need to be done to get this up to shape?

I am still a bit uncertain about the `#![no_std]` in https://github.com/NixOS/nixpkgs/pull/215408/commits/4cfc9ed1b1bc4dcd1ddc47fc0d387d8f94f2ef39 - the sysroot is used for core, alloc, compiler_builtins but not std, right?

**Comment by astro** [comment]
I think backporting would fix 22.11 but the Cargo.lock will probably differ. How do I PR that?

**Comment by figsoda** [comment]
Cherry-picking https://github.com/NixOS/nixpkgs/pull/215408/commits/d284e8dc30e5006d5f1e20b9efd406d4d9bdb557 is what I would do

**Comment by Ericson2314** [comment]
BTW for reference https://github.com/alamgu/alamgu/tree/develop/stdlib is more up to date than what is in Nixpkgs today. I would love to deduplicate that with this. Note that unlike Nixpkgs, I can use the IFD features of `crate2nix` for that.

**Comment by winterqt** [comment]
@Ericson2314 Why/how has sysroot regressed over the past <whatever timeframe>? Should we be adding this to rustc's `passthru.tests` to avoid this in the future? (I admittedly don't know too much about this and its history, apologies.)

**Comment by Ericson2314** [comment]
@winterqt I am not totally sure myself but I think an issue is that is baked from a single `rustc`, so we aren't even the supporting the multiple `rustc`s we may have in Nixpkgs at a time correctly.

I think ideally yes we would have the test, but we would also have the IFD part as some sort of maintainer script that can be run to regenerate the thing as needed.

**Comment by winterqt** [comment]
> @winterqt I am not totally sure myself but I think an issue is that is baked from a single `rustc`, so we aren't even the supporting the multiple `rustc`s we may have in Nixpkgs at a time correctly.

I'd say it'd really only make sense to support the latest in Nixpkgs, as we never really have any other versions (except if we need to fix regressions, but even then...)

> I think ideally yes we would have the test, but we would also have the IFD part as some sort of maintainer script that can be run to regenerate the thing as needed.

Do we do IFD in-tree or are you talking about your copy? (I can't find IFD in either.) What would this do over the `update-lockfile.sh` script -- as in, what needs IFD?

**Comment by Ericson2314** [comment]
> I'd say it'd really only make sense to support the latest in Nixpkgs, as we never really have any other versions (except if we need to fix regressions, but even then...)

OK fair enough, I remember there being multiple versions, but I must be out of date.

> Do we do IFD in-tree or are you talking about your copy? (I can't find IFD in either.)

Sorry, it is not trivial find. The *caller* of the stuff I linked, https://github.com/alamgu/alamgu/blob/develop/per-device.nix#L113-L121 does the IFD.

> What would this do over the `update-lockfile.sh` script -- as in, what needs IFD?

This is getting into https://github.com/NixOS/rfcs/pull/109/files, but basically I feel that maintainer scripts that are better sandboxed are more reliable. My memory is hazy at this point, but I recall when I vendored this stuff not only was the Nixpkgs stuff out of date, but the update script was not working.

The update script I envision today *wouldn't* do IFD, but it *would* call `crate2nix` inside a derivation (`crate2nix-tools.generatedCargoNix ...`). It would then copy the result of that into Nixpkgs. This ensures as much of process is pure/reproduciable as possible. Does that make sense?

**Comment by winterqt** [comment]
> This is getting into https://github.com/NixOS/rfcs/pull/109/files, but basically I feel that maintainer scripts that are better sandboxed are more reliable.

I see. (Was there anything in particular that made you mention this? e.g. the current update script in-tree isn't necessarily *bad*...)

> The update script I envision today *wouldn't* do IFD, but it *would* call `crate2nix` inside a derivation (`crate2nix-tools.generatedCargoNix ...`). It would then copy the result of that into Nixpkgs. This ensures as much of process is pure/reproduciable as possible. Does that make sense?

To be clear, by mentioning IFD I was also envisioning just replacing that process by doing whatever to generate the expression and then just vendoring it.

**Comment by Ericson2314** [comment]
> (Was there anything in particular that made you mention this? e.g. the current update script in-tree isn't necessarily bad...)

I am just mentioning this stuff because I had been meaning to fix these things in Nixpkgs based on that repo before you beat me to it. :) Not in response to the concrete changes so you've made so far!

**Comment by winterqt** [comment]
> I am just mentioning this stuff because I had been meaning to fix these things in Nixpkgs based on that repo before you beat me to it. :) Not in response to the concrete changes so you've made so far!

FWIW: I'm not the author, just a reviewer :D I think my questions might still stand though? Not sure.

**Comment by Ericson2314** [comment]
Oh right, whoops! Well I'll go look over what's there now vs what I have and then have more concrete things to say.

**Comment by Ericson2314** [comment]
Also whoops I didn't actually change the upgrade script either. There is`Cargo.nix` stuff in Nixpkgs yet and that would be separate.

**Comment by astro** [comment]
I managed to build [](https://github.com/cloud-hypervisor/rust-hypervisor-firmware) for a custom `x86_64-unknown-none` target with these changes. I am removing the *[Draft]* now.

**Comment by Ericson2314** [comment]
It looks good to me!

**Comment by astro** [comment]
Reviews will be appreciated

**Review by figsoda — APPROVED** [review]
don't know too much about custom targets, nothing looks out of the ordinary

File: pkgs/build-support/rust/hooks/default.nix [hunk 2]
```diff
@@ -20,10 +20,20 @@ let
   shortTarget = if targetIsJSON then
       (lib.removeSuffix ".json" (builtins.baseNameOf "${target}"))
     else target;
-  ccForBuild = "${buildPackages.stdenv.cc}/bin/${buildPackages.stdenv.cc.targetPrefix}cc";
-  cxxForBuild = "${buildPackages.stdenv.cc}/bin/${buildPackages.stdenv.cc.targetPrefix}c++";
-  ccForHost = "${stdenv.cc}/bin/${stdenv.cc.targetPrefix}cc";
-  cxxForHost = "${stdenv.cc}/bin/${stdenv.cc.targetPrefix}c++";
+  makeCc = stdenv: lang:
```
[the remark this answers is not in the document]
Comment (astro) [reply pkgs/build-support/rust/hooks/default.nix:23]: The main reason for this change is the check on `stdenv.cc.isClang`. Feel free to suggest readability improvements!

[the remark this answers is not in the document]
Comment (figsoda) [reply pkgs/build-support/rust/hooks/default.nix:23]: if I'm not mistaken, clang should also have `${stdenv.cc.targetPrefix}cc` and `${stdenv.cc.targetPrefix}c++`

[the remark this answers is not in the document]
Comment (astro) [reply pkgs/build-support/rust/hooks/default.nix]: Turns out I can build with gcc so I'm dropping this commit.

**Review by figsoda — APPROVED** [review]

File: pkgs/build-support/rust/hooks/default.nix (hunk 2 above)
[the remark this answers is not in the document]
Comment (astro) [reply pkgs/build-support/rust/hooks/default.nix]: I just remembered the real reason: I was tumbling over [some compiler/linker detection logic in rustc](https://github.com/rust-lang/rust/blob/cae3c936eb963cedc6523baeeaa9a58d13fa4c2e/compiler/rustc_codegen_ssa/src/back/link.rs#L1232) that passed the wrong parameters to lld.

I'm leaving the commit out unless someone else can test it successfully.

**Comment by astro** [comment]
@winterqt @zowoq Would you mind taking another look at this?

**Pull request merged**